### PR TITLE
[Core] Child processes cleaned up by workers if raylet dies

### LINKED ISSUE #35842: [core] Kill worker children process even when raylet exits.
### What happened + What you expected to happen

Right now, we have policy that when the worker exits, it'll kill all its children otherwise there might be some leak.

But one corner case is that if the raylet died, the worker won't kill itself, but instead it'll just exit which could lead to the worker leak.

There are two cases here:
- Worker crashed
  - which should fail the children too, but is also not easy to fix and needs some os support for it.
- Worker can exit by itself
  - which we should cover because we have control of this.

### Versions / Dependencies

master

### Reproduction script

Refer to this PR https://github.com/ray-project/ray/pull/35840

### Issue Severity

None

### PR DESCRIPTION
**Pull request by cadedaniel**
### Summary
This PR has CoreWorkers kill their child processes when they notice the raylet has died. This fixes an edge case missing from https://github.com/ray-project/ray/pull/33976.

### Details
The CoreWorker processes have a periodic check that validates the raylet process is alive. If the raylet is no longer alive, then the CoreWorker processes invoke `QuickExit`. This PR modifies this behavior so that before calling `QuickExit`, the CoreWorker processes kill all child processes.

Note that this does not address the case where CoreWorker processes crash or are killed with SIGKILL. That condition requires the subreaper work proposed in https://github.com/ray-project/ray/issues/34125 / https://github.com/ray-project/ray/issues/26118.

The `ray stop --force` case also requires the subreaper work. This is because `--force` kills CoreWorker processes with SIGKILL, leaving them no opportunity to clean up child processes. We can add the logic which catches leaked child processes, but the best design for this is the subreaper design.

### Issue

Closes https://github.com/ray-project/ray/issues/35842

### CODE REVIEW

File: src/ray/util/process.cc [hunk 1]
```diff
@@ -629,6 +629,8 @@ bool IsProcessAlive(pid_t pid) {
   }
   return false;
 #else
+  // Note if the process is a zombie (dead but not yet reaped), it will
+  // still be alive by this check.
```
Comment (rkooo567) [inline src/ray/util/process.cc:633]: We still kill them inside KillChildProcs right?

**Review by rkooo567 — APPROVED** [review]
LGTM

File: src/ray/util/process.cc (hunk 1 above)
Comment (cadedaniel) [reply src/ray/util/process.cc:633]: Yep, all child procs are sent kill signal regardless of process state

**Comment by rkooo567** [comment]
there are lots of windows failures, but they shouldn't be related (this code doesn't impact windows)

**Pull request merged**